core/src/main/java/org/opensearch/sql/expression/aggregation/AvgAggregator.java

```diff
@@ -47,18 +75,117 @@ public String toString() {
   /**
    * Average State.
    */
-  protected static class AvgState implements AggregationState {
-    private int count;
-    private double total;
+  protected abstract static class AvgState implements AggregationState {
+    protected ExprValue count;
+    protected ExprValue total;
 
     AvgState() {
-      this.count = 0;
-      this.total = 0d;
+      this.count = new ExprIntegerValue(0);
+      this.total = new ExprDoubleValue(0D);
+    }
+
+    @Override
+    public abstract ExprValue result();
+
+    protected AvgState iterate(ExprValue value) {
+      count = DSL.add(DSL.literal(count), DSL.literal(1)).valueOf();
+      return this;
+    }
+  }
+
+  protected static class DoubleAvgState extends AvgState {
+    @Override
+    public ExprValue result() {
+      if (0 == count.integerValue()) {
+        return ExprNullValue.of();
+      }
+      return DSL.divide(DSL.literal(total), DSL.literal(count)).valueOf();
+    }
+
+    @Override
+    protected AvgState iterate(ExprValue value) {
+      total = DSL.add(DSL.literal(total), DSL.literal(value)).valueOf();
+      return super.iterate(value);
+    }
+  }
+
+  protected static class DateAvgState extends AvgState {
+    @Override
+    public ExprValue result() {
+      if (0 == count.integerValue()) {
+        return ExprNullValue.of();
+      }
+
+      return new ExprDateValue(
+          new ExprTimestampValue(Instant.ofEpochMilli(
+              DSL.divide(DSL.literal(total), DSL.literal(count)).valueOf().longValue()))
+                  .dateValue());
+    }
+
+    @Override
+    protected AvgState iterate(ExprValue value) {
```

Comment:
is it possible to reuse `DSL.adddate()`?

Reply:
Yes, but `DateAvgState` will be less clear and readable:
```java
  protected static class DateAvgState extends AvgState {

    public DateAvgState() {
      this.count = new ExprIntegerValue(0);
      this.total = new ExprDateValue(LocalDate.EPOCH);
    }

    @Override
    public ExprValue result() {
      if (0 == count.integerValue()) {
        return ExprNullValue.of();
      }

      return DSL.adddate(DSL.literal(new ExprDateValue(LocalDate.EPOCH)), 
          DSL.literal(DSL.divide(DSL.literal(DAYS.between(LocalDate.EPOCH, total.dateValue())),
              DSL.literal(count)).valueOf().longValue())).valueOf();
    }

    @Override
    protected AvgState iterate(ExprValue value) {
      total = DSL.adddate(DSL.literal(total),
          DSL.literal(DAYS.between(LocalDate.EPOCH, value.dateValue())))
          .valueOf();

      return super.iterate(value);
    }
  }
```